src/runtime_src/core/tools/common/Report.cpp

```diff
@@ -0,0 +1,109 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include "Report.h"
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <sstream>
+
+// Initialize our static mapping.
+const Report::SchemaDescriptionVector Report::m_schemaVersionVector = {
+  { SchemaVersion::UNKNOWN,       false, "",              "Unknown entry"},
+  { SchemaVersion::TEXT,          true,  "text",          "Human readable report"},
+  { SchemaVersion::JSON_INTERNAL, false, "JSON-internal", "Internal JSON property tree"},
+  { SchemaVersion::JSON_20201,    true,  "JSON-2020.1",   "JSON 2020.1 schema"}
+};
+
+
+const Report::SchemaDescription &
+Report::getSchemaDescription(const std::string & _schemaVersionName) 
+{
+  // Look for a match
+  for (const auto & entry : m_schemaVersionVector) {
+    if (boost::iequals(entry.optionName, _schemaVersionName)) 
+      return entry;
+  }
+
+  // Return back the unknown entry
+  return getSchemaDescription("");
+}
+
+const Report::SchemaDescription & 
+Report::getSchemaDescription(SchemaVersion _schemaVersion)
+{
+  // Look for a match
+  for (const auto & entry : m_schemaVersionVector) {
+    if (entry.schemaVersion == _schemaVersion) 
+      return entry;
+  }
+
+  // Return back the unknown entry
+  return getSchemaDescription(SchemaVersion::UNKNOWN);
+}
+
+
+Report::Report(const std::string & _reportName,
+               const std::string & _shortDescription,
+               bool _isDeviceRequired)
+  : m_reportName(_reportName)
+  , m_shortDescription(_shortDescription)
+  , m_isDeviceRequired(_isDeviceRequired)
+{
+  // Empty
+}
+
+
+boost::any 
+Report::getFormattedReport(const xrt_core::device *_pDevice, 
+                           SchemaVersion _schemaVersion,
+                           const std::vector<std::string> & _elementFilter) const
+{
+  // Helper variables
+  boost::property_tree::ptree _pt;
+  std::stringstream _ostream;
+  boost::any returnValue;
+
+  switch (_schemaVersion) {
+    case SchemaVersion::TEXT:  
+      writeReport(_pDevice, _elementFilter, _ostream);
+      returnValue = _ostream.str();
+      return returnValue;
+
+    case SchemaVersion::JSON_INTERNAL:
+      getPropertyTreeInternal(_pDevice, _pt);
+      returnValue = _pt;
+      return returnValue;
+
+    case SchemaVersion::JSON_20201:
+      getPropertyTree20201(_pDevice, _pt);
+      returnValue = _pt;
+      return returnValue;
+
+    case SchemaVersion::UNKNOWN:
+      throw std::runtime_error("ERROR: Unknown schema version.");
+
+    // Note: There is no default in this switch statement.  Relying on the
+    //       compiler to produce a warning indicating that all of the enum values
+    //       are present.
+  }
+
+  // Code will never get here, but the compiler doesn't know that.
+  // Return back an empty value.
+  return boost::any();

```

Comment:
Better to throw std::runtime_error to indicated unexpected error, that will silence the compiler.

Reply:
Made changes as suggested and learned something new today as well.